clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -0,0 +1,484 @@
+static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+  ASTContext &Ctx = Src->getASTContext();
+  std::string Name = (Twine("_arg_") + Src->getName()).str();
+  return std::make_tuple(Ty, &Ctx.Idents.get(Name),
+                         Ctx.getTrivialTypeSourceInfo(Ty));
+}
+
+// Creates list of kernel parameters descriptors using KernelObj (kernel
+// object). Fields of kernel object must be initialized with SYCL kernel
+// arguments so in the following function we extract types of kernel object
+// fields and add it to the array with kernel parameters descriptors. Returns
+// true if all arguments are successfully built.
+static bool buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
+                        SmallVectorImpl<ParamDesc> &ParamDescs) {
+  auto CreateAndAddPrmDsc = [&](const FieldDecl *Fld, const QualType &ArgType) {
+    // Create a parameter descriptor and append it to the result
+    ParamDescs.push_back(makeParamDesc(Fld, ArgType));
+  };
+
+  // Creates a parameter descriptor for SYCL special object - SYCL accessor or
+  // sampler.
+  // All special SYCL objects must have __init method. We extract types for
+  // kernel parameters from __init method parameters. We will use __init method
+  // and kernel parameters which we build here to initialize special objects in
+  // the kernel body.
+  auto createSpecialSYCLObjParamDesc = [&](const FieldDecl *Fld,
+                                           const QualType &ArgTy) {
+    const auto *RecordDecl = ArgTy->getAsCXXRecordDecl();
+    assert(RecordDecl && "Special SYCL object must be of a record type");
+
+    CXXMethodDecl *InitMethod = getInitMethod(RecordDecl);
+    assert(InitMethod && "The accessor/sampler must have the __init method");
+    unsigned NumParams = InitMethod->getNumParams();
+    for (size_t I = 0; I < NumParams; ++I) {
+      ParmVarDecl *PD = InitMethod->getParamDecl(I);
+      CreateAndAddPrmDsc(Fld, PD->getType().getCanonicalType());
+    }
+  };
+
+  bool AllArgsAreValid = true;
+  // Run through kernel object fields and create corresponding kernel
+  // parameters descriptors. There are a several possible cases:
+  //   - Kernel object field is a SYCL special object (SYCL accessor or SYCL
+  //     sampler). These objects has a special initialization scheme - using
+  //     __init method.
+  //   - Kernel object field has a scalar type. In this case we should add
+  //     kernel parameter with the same type.
+  //   - Kernel object field has a structure or class type. Same handling as a
+  //     scalar but we should check if this structure/class contains accessors
+  //     and add parameter decriptor for them properly.
+  for (const auto *Fld : KernelObj->fields()) {
+    QualType ArgTy = Fld->getType();
+    if (Util::isSyclAccessorType(ArgTy)) {
+      createSpecialSYCLObjParamDesc(Fld, ArgTy);
+    } else if (!ArgTy->isStandardLayoutType()) {
+
+      // TODO: Do we need diagnostics with the kernel gen patch?
```

Comment:
@bader , what do you think about adding diagnostics with OpenCL kernel generation patch?

Reply:
Please, remove and proceed with uploading the patch to reviews.llvm.org.